File: export/get_citibike_status.py
""" Pulls current station statuses from Citibike API and inserts them into mongo """
import requests
import json
import sys
from pymongo import MongoClient
from datetime import datetime
from apscheduler.schedulers.background import BackgroundScheduler
from time import sleep

client = MongoClient('ds055709.mongolab.com',55709)
db = client['citibike']
db.authenticate('admin','s10thr0p')
collection = db['stationstatuses']


def load_station_statuses():
    response = requests.get(
        'http://appservices.citibikenyc.com/data2/stations.php?updateOnly=true')
    response_json = json.loads(response.text)

    last_update_int = int(response_json['lastUpdate'])
    d = datetime.fromtimestamp(last_update_int)

    results = response_json['results']

    test = collection.find_one({'year': d.year, 'month': d.month, 'day': d.day, 'hour': d.hour, 'minute': d.minute})

    print(test)

    if(test is None):
        for result in results:
            status = {
                'bikes': result['availableBikes'],
                'docks': result['availableDocks'],
                'status': result['status'],
                'year': d.year,
                'month': d.month,
                'day': d.day,
                'weekday': d.weekday(),
                'hour': d.hour,
                'minute': d.minute,
                'date': d
            }
            collection.insert(status)
            print(status)


def main():
    sched = BackgroundScheduler()
    job = sched.add_job(load_station_statuses, 'interval', minutes=10)
    sched.start()


if __name__ == '__main__':
    main()
    while(True):
        sleep(1)
        sys.stdout.write('.')
        sys.stdout.flush()

File: export/get_citibike_status.js
var request = require('request');
var mongoose = require('mongoose');
var schedule = require('node-schedule');
// var stationStatus = require('StationStatusDB');

mongoose.connect('mongodb://localhost/citibike');

var StationStatus = mongoose.model('StationStatus', { 
    'date': Date, 
    'id': Number, 
    'status':String,
    'bikes': Number, 
    'docks': Number, 
    'year':Number, 
    'month':Number, 
    'day':Number, 
    'weekday':Number,
    'hour':Number,
    'minute':Number
});

var options = {
    url: 'http://appservices.citibikenyc.com/data2/stations.php?updateOnly=true',
    json: true,
    headers: {
        'User-Agent': 'request'
    }
};

var station_status_job = schedule.scheduleJob('0,15,30,45 * * * *', function() {
    request(options, function (error, response, body) {
        if (!error && response.statusCode == 200) {
            var lastupdate = body.lastUpdate;
            var results = body.results;
            var date = new Date(Number(lastupdate)*1000);
            
            var query = StationStatus.findOne({'year':date.getFullYear(),'month':date.getMonth(),'day':date.getDate(),'hour':date.getHours(),'minute':date.getMinutes()});
            query.exec(function(err, existing_record) {
                console.log('HAS EXISTING: ' + (existing_record != null)); 
                if(existing_record == null) { 
                    console.log('SAVING ' + results.length + 'records for ' + date);
                    results.forEach(function(result) {
                        var stationStatus = new StationStatus({
                            'date':date, 
                            'id':result.id, 
                            'status':result.status,
                            'bikes':result.availableBikes, 
                            'docks':result.availableDocks,
                            'year':date.getFullYear(),
                            'month':date.getMonth(),
                            'day':date.getDate(),
                            'weekday':date.getDay(),
                            'hour':date.getHours(),
                            'minute':date.getMinutes()
                        });
                        stationStatus.save(function(err) {
                            if(err) {
                                console.log('ERROR: ' + err);
                            } else {
                                console.log('SAVED: ' + stationStatus.id + '|' + stationStatus.bikes + '|' + stationStatus.docks);
                            }
                        });
                    });
                };
            });
        };
    });
});